lib/docassemble/docassemble_api_interface.js

```diff
@@ -0,0 +1,377 @@
+  * let { time_per_loop, max_loops } = options || { time_per_loop: 5 * 1000, max_loops: 15 };
+  */
+
+  // Defaults
+  let { timeout } = options || { timeout: 5 * 1000 };
+  log.info({ pre: `Waiting for the docassemble server to respond. Will wait for ${ timeout/1000 } seconds at most.` });
+  
+  try {
+    let start_time = Date.now()
+    // Axios + server will poll until server is ready or timeout is over
+    let response = await _da.get_dev_id( timeout );
+    let end_time = Date.now()
+    if ( response && response.data && response.data.id ) {
+      log.info({ pre: `The docassemble server responded after ${ (end_time - start_time)/1000 } seconds.` });
+      return false;
+    }
+
+  } catch ( error ) {
+    // Timeout means the server is busy.
+    if ( is_timeout_error( error ) ) {
+      // Anything that's waiting for the server to not be busy does not want
+      // to move forward without a free server.
+      log.info({ pre: `The docassemble server was still busy after ${ timeout/1000 } seconds.` });
+    } else {
+      // Not sure what to do for other errors. It could be a server
+      // error, like an 'Access Denied' error, which should be handled.
+      log.warn({ pre: `Tried to check if the docassemble server was busy, but ran into an unexpected `
+        + `error. The error didn't indicate a busy server, though.`
+      });
+    }
+
+    throw error;
+  }
+};  // Ends da_i.throw_an_error_if_server_is_not_responding()
+
+
+// node -e 'require("./lib/github_integration/docassemble_api_interface.js").get_dev_id()'
+da_i.get_dev_id = async function ( dev_id_options ) {
+  /** For consistency, wrap _da.get_dev_id().
+  *    See https://docassemble.org/docs/api.html#user.
+  *    Should we throw the 'Access Denied' error in here instead? */
+
+  // Default timeouts
+  let { timeout } = dev_id_options ||  { timeout: 75 * 1000 };
+  return await _da.get_dev_id( timeout );
+};  // Ends da_i.get_dev_id();
+
+
+// node -e 'require("./lib/github_integration/docassemble_api_interface.js").try_api_key()'
+da_i.try_api_key = async function ( api_key_options ) {
+  /** Make sure the api key is the right one for this server.
+  *   See https://docassemble.org/docs/api.html#user. */
+
+  log.info({ pre: `Checking that the API key of the docassemble server's testing account is valid.` });
+
+  // Default timeouts
+  api_key_options = api_key_options ||  { timeout: 75 * 1000 };
+
+  try {
+
+    // There's only one possible success result
+    // As long as this succeeds, we're good
+    await da_i.get_dev_id( api_key_options );
+    log.info({ pre: `Your testing account's API key is perfectly valid on this server.` });
+    return true;
+
+  } catch ( error ) {
+    // There's only one possible sever response error
+    if ( error.response && error.response.status === 403 ) {
+      // Throw in here, or at the caller
+      throw( `Error: Request failed with status code 403, "Access `
```

Comment:
Is there a more specific exception type that we could throw here, or is that not how node works?  In Python, for example, you might want to throw a requests.AccessDeniedException during an AccessDenied situation.

Reply:
Errors (exceptions) in js do have type names, including server response errors. I don't think any "local" error types fit here. I think it's most appropriate as a server response error. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#error_types.

I also think that sometimes those error type names aren't helpful. E.g. 403's type is "forbidden". I think that would be more alarming than helpful. My take for this situation - the message here gives much more information and if they want to find out more about the error type, looking up "403" would be enough. Let me know what you think! See https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403.

Regardless, we should definitely watch out for other places where these type names would be helpful.

Reply:
If you want to include a type, you can do it this way:
```js
throw new ReferenceError( `Text of error` );
```